### PR TITLE
PC-1220: Connect to Open EPC API

### PR DESCRIPTION
**Pull request by samyou-softwire**
# Description

this is draft as [PC-1299](https://beisdigital.atlassian.net/browse/PC-1299) is not yet merged

# Checklist

- [ ] I have made any necessary updates to the documentation
- [ ] I have checked there are no unnecessary IDE warnings in this PR
- [ ] I have checked there are no unintentional line ending changes
- [ ] I have added tests where applicable
- [ ] If I have made any changes to the code, I have used the IDE auto-formatter on it
- [ ] If I have made any changes to Python files (even if not changing any content strings), I have run `make extract-translations`
- [ ] If I have made any changes to website flow, I have updated the [Flow Miro Board](https://miro.com/app/board/uXjVK4AgQ3E=/)
- [ ] If I have made any changes to website flow, I have checked forward and back behaviour is still consistent

# Screenshots

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ Formerly known as "Help to Heat".
 ## Initial Setup
 1. From the project root, run `cp envs/web.template envs/web`
 2. Populate the OS API key into the OS_API_KEY variable in `envs/web`. The key can be found in Keeper. Please only use a single value (i.e. `["my_key_here"]`), rather than all of the keys that are stored in Keeper, in order to avoid needing to rotate all of the keys if your local environment is accidentally leaked.
+3. Populate the Open EPC API token into the OPEN_EPC_API_TOKEN variable in `envs/web`. This key can be found in keeper.
```
Comment (jdgage) [inline README.md:7]: Applies to the above as well, but I suppose the bit about Keeper doesn't belong here as it's Softwire-specific.

File: help_to_heat/frontdoor/epc_api.py [hunk 2]
```diff
@@ -8,35 +8,25 @@
...
-
-        response = requests.post(token_url, data=payload)
-        response.raise_for_status()
-        return response.json().get("access_token")
-
-    def get_address_and_rrn(self, building, postcode):
-        base_url = settings.EPC_API_BASE_URL
-        params = urllib.parse.urlencode({"postcode": postcode, "buildingNameOrNumber": building})
-        url = f"{base_url}/api/assessments/domestic-epcs/search?{params}"
+    def get_address_and_lmk(self, building, postcode):
```
Comment (jdgage) [inline help_to_heat/frontdoor/epc_api.py:14]: What does this actually return? Is it more than just address and LMK?

File: help_to_heat/frontdoor/epc_api.py [hunk 3]
```diff
@@ -8,35 +8,25 @@
...
+        base_url = settings.OPEN_EPC_API_BASE_URL
+        params = urllib.parse.urlencode({"postcode": postcode, "address": building})
+        url = f"{base_url}/search?{params}"
         return self.__api_call(url)
 
-    def get_epc_details(self, rrn):
-        base_url = settings.EPC_API_BASE_URL
-        rrn_for_path = urllib.parse.quote(rrn)
-        url = f"{base_url}/api/ecoplus/assessments/{rrn_for_path}"
+    def get_epc_details(self, lmk):
```
Comment (jdgage) [inline help_to_heat/frontdoor/epc_api.py:20]: Likewise, what shape of thing does this return?

File: help_to_heat/frontdoor/interface.py [hunk 4]
```diff
@@ -534,53 +534,19 @@ def get_epc_scotland(self, uprn):
...
-            return None
-
-    def __save_token(self, new_token):
-        saved_token = models.AccessToken.objects.first()
-        if saved_token is not None:
-            saved_token.access_token = new_token
-        else:
-            saved_token = models.AccessToken(access_token=new_token)
-        saved_token.save()
+            raise e
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:549]: Is there any benefit to catching if we're going to re-throw straight away? (and then is there any benefit to this wrapper method at all?)

**Comment by jdgage** [comment]
Broadly looks fine: couple of comments. Main thing to think about is how we document what we're expecting to get back from the API.

File: help_to_heat/frontdoor/interface.py (hunk 4 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py:549]: perhaps the wrapper can stay as it means it's only needed to construct `EPCApi` in one place, perhaps could be renamed

agree the try isn't much needed

File: help_to_heat/frontdoor/epc_api.py (hunk 2 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/epc_api.py:14]: it does return the whole search. can update the method name to reflect this & point to the example json object we have in the project

File: README.md (hunk 1 above)
Comment (samyou-softwire) [reply README.md:7]: have removed

File: help_to_heat/frontdoor/interface.py (hunk 4 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py]: left the wrapper method for now

Comment (jdgage) [reply help_to_heat/frontdoor/interface.py]: Not totally confident in my understanding of patching in Python, but I don't immediately see why the tests wouldn't wok with e.g.
```
api = EpcApi()
data = api.get_address_and_lmk(building_name_or_number, postcode)
```
? It should be equivalent. Would be quite a lot clearer than what we have at the moment - feels like machinery for the sake of it.

**Review by jdgage — APPROVED** [review]

**Pull request merged**